subsys/bluetooth/controller/hci/hci.c

```diff
@@ -4798,6 +4798,95 @@ static void vs_read_tx_power_level(struct net_buf *buf, struct net_buf **evt)
 	rp->handle = sys_cpu_to_le16(handle);
 }
 #endif /* CONFIG_BT_CTLR_TX_PWR_DYNAMIC_CONTROL */
+
+#if defined(CONFIG_BT_HCI_VS_FATAL_ERROR)
+/* A memory pool for vandor specific events for fatal error rerporting purposes. */
+NET_BUF_POOL_FIXED_DEFINE(vs_err_tx_pool, 1, BT_BUF_RX_SIZE, 8, NULL);
+static struct net_buf *vs_err_evt_create(uint8_t subevt, uint8_t len)
+{
+	struct net_buf *buf;
+
+	buf = net_buf_alloc(&vs_err_tx_pool, K_FOREVER);
+	if (buf) {
+		struct bt_hci_evt_le_meta_event *me;
+		struct bt_hci_evt_hdr *hdr;
+
+		net_buf_reserve(buf, BT_BUF_RESERVE);
+		bt_buf_set_type(buf, BT_BUF_EVT);
+
+		hdr = net_buf_add(buf, sizeof(*hdr));
+		hdr->evt = BT_HCI_EVT_VENDOR;
+		hdr->len = len + sizeof(*me);
+
+		me = net_buf_add(buf, sizeof(*me));
+		me->subevent = subevt;
+	}
+
+	return buf;
+}
+
+struct net_buf *hci_vs_err_fatal(unsigned int reason, const z_arch_esf_t *esf)
+{
+	/* Prepare vendor specific HCI Fatal Error event */
+	struct bt_hci_vs_fatal_error_stack_frame *sf;
+	struct net_buf *buf;
+
+	buf = vs_err_evt_create(BT_HCI_EVT_VS_ERROR_TYPE_SYSTEM, sizeof(*sf));
+	BT_ASSERT_MSG(buf != NULL, "Fatal error\n");
+
+	sf = net_buf_add(buf, sizeof(*sf));
+	sf->reason = sys_cpu_to_le32(reason);
+#if defined(CONFIG_CPU_CORTEX_M)
+	sf->a1 = sys_cpu_to_le32(esf->basic.a1);
+	sf->a2 = sys_cpu_to_le32(esf->basic.a2);
+	sf->a3 = sys_cpu_to_le32(esf->basic.a3);
+	sf->a4 = sys_cpu_to_le32(esf->basic.a4);
+	sf->ip = sys_cpu_to_le32(esf->basic.ip);
+	sf->lr = sys_cpu_to_le32(esf->basic.lr);
+	sf->xpsr = sys_cpu_to_le32(esf->basic.xpsr);
+#endif /* CONFIG_CPU_CORTEX_M */
+
+	return buf;
+}
+
+struct net_buf *hci_vs_err_assert(char *file, uint32_t line)
+{
+	uint32_t len = 0U, pos = 0U;
+	struct net_buf *buf;
+
+	/* Calculate length of a string holding source file name */
+	if (file) {
+		while (file[len] != '\0') {
+			if (file[len] == '/') {
+				pos = len + 1;
+			}
+			len++;
+		}
+		file += pos;
+		len -= pos;
+	}
+
+	/* Total length: fine name strlen + \0 + 32-bit line number */
+	len = 1 + len + 1 + 4;
```

Comment:
So the last 1 and the 4 is the `\0` and the 32-bit line number, and I guess the `len` is the name (do you mean `file` and not `fine`?), but what is the first `1`?

Reply:
Yes it should be `file` and first `1` is a leftover.

Reply:
Corrected